db/src/Pos/DB/Update/Logic/Global.hs

```diff
@@ -186,16 +186,49 @@ usVerifyBlocks genesisConfig verifyAllIsKnown blocks = do
     processRes (Left failure, _)       = Left failure
     processRes (Right undos, modifier) = Right (modifier, undos)
 
-verifyBlock
-    :: (USGlobalVerifyMode ctx m, MonadPoll m, MonadError PollVerFailure m)
+verifyBlock ::
+    ( USGlobalVerifyMode ctx m
+    , DB.MonadDBRead m
+    , MonadPoll m
+    , MonadError PollVerFailure m
+    )
     => Genesis.Config
-    -> BlockVersion
     -> Bool
     -> UpdateBlock
     -> m USUndo
-verifyBlock genesisConfig _ _ (ComponentBlockGenesis genBlk) =
+verifyBlock genesisConfig _ (ComponentBlockGenesis genBlk) =
     execRollT $ processGenesisBlock genesisConfig (genBlk ^. epochIndexL)
-verifyBlock genesisConfig lastAdopted verifyAllIsKnown (ComponentBlockMain header payload) =
+verifyBlock genesisConfig verifyAllIsKnown (ComponentBlockMain header payload) = do
+    lastAdopted <- do
+        -- During the `Original` era, `processGenesisBlock` is typically
+        -- called from `verifyBlock` when provided with a
+        -- `ComponentBlockGenesis` argument.
+        --
+        -- `processGenesisBlock` is responsible for some important
+        -- operations which are to occur at the epoch boundary such as
+        -- handling update adoption.
+        --
+        -- Since epoch boundary blocks aren't created during OBFT,
+        -- we must ensure that we call `processGenesisBlock` on epoch
+        -- boundaries (first slot of a new epoch) such that updates can still
+        -- be adopted during OBFT.
+        initialEra <- getConsensusEra
+        initialBV  <- getAdoptedBV
+        logDebug $ "usVerifyBlocks: era '" <> show initialEra <> "'"
+        case initialEra of
+            OBFT _ -> do
+                logDebug $ "usVerifyBlocks OBFT: Checking whether we're"
+                    <> " on epoch boundary and should attempt update"
+                let slotId     = header ^. headerSlotL
+                    epochIndex = siEpoch slotId
+                tipHeader <- DB.getTipHeader
+                whenEpochBoundaryObft epochIndex tipHeader (\ei -> do
+                    logDebug $ "usVerifyBlocks OBFT: We're on epoch boundary. Running processGenesisBlock"
+                    processGenesisBlock genesisConfig ei)
+                getAdoptedBV
+
+            Original -> pure initialBV
```

Comment:
👍 I like the refactor here to drop the `BlockVersion` arg and instead check `getAdoptedBV` inside of `verifyBlock`, so that we can catch updates as we go along.